File: day02/day02.go
package day02

import (
	"fmt"
	"os"
	"strings"
)

const (
	ROCK = iota
	PAPER
	SCISSORS
)

func getMove(move byte) byte {
	if move < 'D' {
		return move - 'A'
	} else {
		return move - 'X'
	}
}

func calcScore(line string) int {
	opponent := getMove(line[0])
	self := getMove(line[2])
	// score for pick
	score := int(self + 1)
	// score for draw
	if self == opponent {
		score += 3
	}
	// score for winning
	if (self+2)%3 == opponent {
		score += 6
	}
	return score
}

func decideMove(opponent, self byte) byte {
	// start with draw move
	move := opponent
	switch self {
	case 'X':
		// pick losing move
		move = (opponent + 2) % 3
	case 'Z':
		// pick winning move
		move = (opponent + 1) % 3
	}
	return move
}

func calcScore2(line string) int {
	opponent := getMove(line[0])
	self := decideMove(opponent, line[2])
	// score for pick
	score := int(self + 1)
	// score for draw
	if self == opponent {
		score += 3
	}
	// score for winning
	if (self+2)%3 == opponent {
		score += 6
	}
	return score
}

func Solve(test bool) (string, string) {
	input, _ := os.ReadFile("day02/input.txt")
	lines := strings.Split(string(input), "\n")
	score := 0
	for _, line := range lines {
		if line == "" {
			continue
		}
		score += calcScore(line)
	}
	fmt.Printf("Score 1: %v\n", score)
	res1 := fmt.Sprintf("%d", score)

	score = 0
	for _, line := range lines {
		if line == "" {
			continue
		}
		score += calcScore2(line)
	}
	fmt.Printf("Score 2: %v\n", score)
	res2 := fmt.Sprintf("%d", score)
	return res1, res2
}

File: aoc22_rust/src/day02.rs
use crate::types::Task;

fn get_move(m: u8) -> u8 {
    if m < b'D' {
        return m - b'A';
    } else {
        return m - b'X';
    }
}

fn calc_score(line: &str) -> i32 {
    let opponent = get_move(line.bytes().nth(0).unwrap());
    let me = get_move(line.bytes().nth(2).unwrap());
    // score for pick
    let mut score = me as i32 + 1;
    // score for draw
    if me == opponent {
        score += 3
    }
    // score for winning
    if (me + 2) % 3 == opponent {
        score += 6
    }
    score
}

fn decide_move(opponent: u8, me: u8) -> u8 {
    match me {
        b'X' => {
            // pick losing move
            (opponent + 2) % 3
        }
        b'Z' => {
            // pick winning move
            (opponent + 1) % 3
        }
        _ => {
            // pick draw move
            opponent
        }
    }
}

fn calc_score2(line: &str) -> i32 {
    let opponent = get_move(line.bytes().nth(0).unwrap());
    let me = decide_move(opponent, line.bytes().nth(2).unwrap());
    // score for pick
    let mut score = me as i32 + 1;
    // score for draw
    if me == opponent {
        score += 3
    }
    // score for winning
    if (me + 2) % 3 == opponent {
        score += 6
    }
    score
}

pub fn solve(input: &str, _debug: bool, task: Task) -> (String, String) {
    let (mut res1, mut res2) = ("".to_string(), "".to_string());
    let lines: Vec<&str> = input.lines().collect();
    if !matches!(task, Task::Two) {
        let mut score = 0;
        for line in &lines {
            if line.len() == 0 {
                continue;
            }
            score += calc_score(line)
        }
        res1 = score.to_string();
    }

    if !matches!(task, Task::One) {
        let mut score = 0;
        for line in &lines {
            if line.len() == 0 {
                continue;
            }
            score += calc_score2(line)
        }
        res2 = score.to_string();
    }
    (res1, res2)
}
